src/python/pants/backend/python/subsystems/pex_build_util.py

```diff
@@ -267,13 +292,52 @@ def add_sources_from(self, tgt):
       raise TaskError('Old-style resources not supported for target {}.  '
                       'Depend on resources() targets instead.'.format(tgt.address.spec))
 
+  def _prepare_inits(self):
+    chroot = self._builder.chroot()
+    sources = chroot.get('source') | chroot.get('resource')
+
+    packages = set()
+    for source in sources:
+      if source.endswith('.py'):
+        pkg_dir = os.path.dirname(source)
+        if pkg_dir and pkg_dir not in packages:
+          package = ''
+          for component in pkg_dir.split(os.sep):
+            package = os.path.join(package, component)
+            packages.add(package)
+
+    missing_pkg_files = []
+    for package in packages:
+      pkg_file = os.path.join(package, '__init__.py')
+      if pkg_file not in sources:
+        missing_pkg_files.append(pkg_file)
+
+    if missing_pkg_files:
+      with temporary_file() as ns_package:
+        ns_package.write(b'__import__("pkg_resources").declare_namespace(__name__)')
+        ns_package.flush()
+        for missing_pkg_file in missing_pkg_files:
+          self._builder.add_source(ns_package.name, missing_pkg_file)
+
+    return missing_pkg_files
+
+  def set_emit_warnings(self, emit_warnings):
+    self._builder.info.emit_warnings = emit_warnings
+
   def freeze(self):
-    self._builder.freeze()
+    if not self._frozen:
+      if self._prepare_inits():
```

Comment:
I think this is better expressed as
```python
if not self._frozen and self._prepare_inits():
```

There's no else clause here so leads to less indentation.

Reply:
Nack - only lines 330-332 apply under both conditions. Lines 333-334 should execute if `not self._frozen` regardless of whether or not `__init__.py`s were prepared.